src/dbnode/persist/fs/commitlog/chunk_reader.go

```diff
@@ -80,18 +83,31 @@ func (r *chunkReader) readHeader() error {
 		return err
 	}
 
-	// Verify data checksum
-	data, err := r.buffer.Peek(int(size))
+	// Setup a chunk data buffer so that chunk data can be loaded into it.
+	chunkDataSize := int(size)
+	if chunkDataSize > cap(r.chunkData) {
```

Comment:
I was exploring this idea as well. But unfortunately len(p) (i.e. length of log entry) is unrelated to DataSize so if len(p) < DataSize then we would not be able to read all the chunk bytes and hence wont be able to compute checksum. 

I agree that we need something better and we need to avoid one extra layer of copying buffer that this commit has introduced. I am thinking of removing buffer io all together and going directly to fd to read data and buffer it in chunkData. This will keep the number of buffer copies same as before. What are your thoughts on this? Also can you point me to a benchmark which I can run to quantify the potential performance degradation introduced by this change.

I believe that commit logs are not read in data read/write path since they are served off of the buckets. As a result, I think we wont notice any perf degradation because of the extra layer of buffer copy introduced by this change.

Reply:
Great explanation, thank you! So I think the thing to do here may just be to avoid perverse allocation situations, e.g.:

```
curBufferSize: 4
writeSize: 5 (expand)
writeSize: 6 (expand)
```

Thoughts on doubling the buffer size instead (a la most array list implementations)? I was hoping we could use some go primitives for that, but I'm not sure they're readily available (append is no good to us here, I don't think).

I think if we avoid the perverse situations we're likely 'good enough'.

Reply:
Agreed. Double the required number of byte should be good enough to prevent reallocation several times.

Reply:
I'm not aware of any commitlog-specific benchmarks unfortunately. Maybe it won't be that complicated to write one?

In the meantime I'll deploy to a staging environment as a sanity test (it'll be hard to verify exact performance/memory impact of this change in staging because the number of writes vary too greatly for that to be an apples to apples comparison).